src/Control/Monad/Free.purs

```diff
@@ -1,125 +1,79 @@
 module Control.Monad.Free
-  ( Free(..), GosubF()
-  , FreeC(..)
-  , MonadFree, wrap
-  , liftF, liftFI, liftFC, liftFCI
-  , pureF, pureFC
-  , mapF, mapFC
-  , bindF, bindFC
-  , injF, injFC
+  ( Free()
+  , suspendF
+  , liftF
+  , liftFI
+  , mapF
+  , injF
+  , foldMapF
   , runFree
   , runFreeM
-  , runFreeC
-  , runFreeCM
   ) where
 
-import Prelude
+import Prelude hiding (append)
 
-import Data.Exists
+import Control.Monad.Rec.Class (MonadRec, tailRecM)
+import Control.Monad.Trans (MonadTrans)
 
-import Control.Monad.Trans
-import Control.Monad.Eff
-import Control.Monad.Rec.Class
-
-import Data.Identity
-import Data.Coyoneda
-import Data.Either
-import Data.Function
-import Data.Maybe
+import Data.Either (Either(..), either)
+import Data.Identity (Identity(..), runIdentity)
 import Data.Inject (Inject, inj)
+import Data.Maybe (Maybe(..))
+import Data.NaturalTransformation (NaturalTransformation())
+import Data.Tuple (Tuple(..))
 
-newtype GosubF f a i = GosubF { a :: Unit -> Free f i, f :: i -> Free f a }
-
-gosub :: forall f a i. (Unit -> Free f i) -> (i -> Free f a) -> Free f a
-gosub a f = Gosub $ mkExists $ GosubF { a: a, f: f}
+import Unsafe.Coerce (unsafeCoerce)
 
--- | The free `Monad` for a `Functor`.
+-- | The free monad for a type constructor `f`.
 -- |
--- | The implementation defers the evaluation of monadic binds so that it
--- | is safe to use monadic tail recursion, for example.
-data Free f a = Pure a
-              | Free (f (Free f a))
-              | Gosub (Exists (GosubF f a))
+-- | Implemented in the spirit of [Relection without Remorse](http://okmij.org/ftp/Haskell/zseq.pdf),
+-- | the free monad is represented using a sequential data structure in
+-- | order to overcome the quadratic complexity of left-associated binds
+-- | and traversal through the free monad structure.
+data Free f a = Free (FreeView f Val Val) (List (ExpF f))
 
--- | The free `Monad` for an arbitrary type constructor.
-type FreeC f = Free (Coyoneda f)
+newtype ExpF f = ExpF (Val -> Free f Val)
 
--- | The `MonadFree` class provides the `wrap` function, which lifts
--- | actions described by a generating functor into a monad.
--- |
--- | The canonical instance of `MonadFree f` is `Free f`.
-class MonadFree f m where
-  wrap :: forall a. f (m a) -> m a
+data FreeView f a b = Return a | Bind (f b) (b -> Free f a)
 
-instance functorFree :: (Functor f) => Functor (Free f) where
-  map f (Pure a) = Pure (f a)
-  map f g = liftA1 f g
+data Val
 
-instance applyFree :: (Functor f) => Apply (Free f) where
-  apply = ap
+instance freeFunctor :: Functor (Free f) where
+  map k f = f >>= return <<< k
 
-instance applicativeFree :: (Functor f) => Applicative (Free f) where
-  pure = Pure
+instance freeBind :: Bind (Free f) where
+  bind (Free v s) k = Free v (snoc s (ExpF (unsafeCoerceBind k)))
```

Comment:
You can take the same Free and bind it once, and also bind it twice, and throw away the latter one. Now the list size for the first one will be wrong. So it's not safe.